src/main/java/sprint06/task04/Main.java

```diff
@@ -0,0 +1,41 @@
+package sprint06.task04;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Main {
+    public static void main(String[] args) {
+        Shop shop = new Shop();
+        shop.clients.add(new Person("person1").goShopping);
+        shop.clients.add(new Person("person2").goShopping);
+        shop.clients.add(new Person("person3").goShopping);
+        shop.clients.add(new Person("person4").goShopping);
+        shop.clients.add(((product, percent) -> percent == 0));
+        System.out.println(shop.sale("product2", 10));
+        System.out.println(shop.sale("product1", 8));
+        System.out.println(shop.sale(null, 0));
+        System.out.println(shop.sale(null, 11));
+    }
+}
+
+class Person {
+    String name;
+
+    Person(String name) {
+        this.name = name;
+    }
+
+    DecisionMethod goShopping = (product, percent) -> product == null ? percent > 10 : product.equals("product1") || percent > 10;
```

Comment:
To avoid excessive null check you can compare from the non-null constant as:
"product1".equals(product)

Reply:
ty)